### ISSUE TITLE
FastDeploy.DeployFileWithFastDevTool failed with "Could not find file "/.../MyApp/obj/Debug/android/assets/Acr.UserDialogs.dll.mdb" error

### ISSUE THREAD

**Issue by alexeystrakh** [xamarin/xamarin-android#5631]
### Steps to Reproduce

1. Create a Xamarin.Forms app
2. Reference Acr.UserDialogs 7.0.35 package (the version is important, the latest version of the package works)
3. Run the Android target on an Android device (or emulator)

### Expected Behavior

The app is built and deployed to the device

### Actual Behavior

The app is built but the deployment step fails with FastDeploy.DeployFileWithFastDevTool failure:

```
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(586,5): error : Could not find file "/Users/my_user/Desktop/Sources/MyApp/Droid/obj/Debug/android/assets/Acr.UserDialogs.dll.mdb"
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(586,5): error :   at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x00259] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.IO/FileStream.cs:274 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(586,5): error :   at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.IO/FileStream.cs:91 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(586,5): error :   at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare)
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(586,5): error :   at System.IO.File.OpenRead (System.String path) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/external/corefx/src/System.IO.FileSystem/src/System/IO/File.cs:266 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(586,5): error :   at Xamarin.Android.Tasks.FastDeploy.DeployFileWithFastDevTool (Microsoft.Build.Framework.ITaskItem file, System.String toolPath, System.String overridePath, System.Byte[] buffer, System.Byte[] compressed, K4os.Compression.LZ4.LZ4Level lz4level) [0x00061] in /Users/builder/azdo/_work/278/s/xamarin-android/external/monodroid/tools/msbuild/Tasks/FastDeploy.cs:594 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(586,5): error :   at Xamarin.Android.Tasks.FastDeploy.DeployFastDevFiles (System.String toolPath, System.String overridePath) [0x0061c] in /Users/builder/azdo/_work/278/s/xamarin-android/external/monodroid/tools/msbuild/Tasks/FastDeploy.cs:573 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(586,5): error :   at Xamarin.Android.Tasks.FastDeploy.RunTaskAsync () [0x00574] in /Users/builder/azdo/_work/278/s/xamarin-android/external/monodroid/tools/msbuild/Tasks/FastDeploy.cs:223 
    4 Warning(s)
    1 Error(s)
```

### Version Information

```
=== Visual Studio Enterprise 2019 for Mac (Preview) ===

Version 8.9 Preview (8.9 build 1593)
Installation UUID: 197b4b8d-2a9a-4fbd-896c-cd74849affd3
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 6.18.0.23 (d16-6 / 088c73638)

	Package version: 612000122

=== Mono Framework MDK ===

Runtime:
	Mono 6.12.0.122 (2020-02/c621c35ffa0) (64-bit)
	Package version: 612000122

=== Roslyn (Language Service) ===

3.9.0-4.21104.8+2fc92d84e746c3aab75f0930278ea6675cd5bb5c

=== NuGet ===

Version: 5.8.0.6860

=== .NET Core SDK ===

SDK: /usr/local/share/dotnet/sdk/5.0.103/Sdks
SDK Versions:
	5.0.103
	5.0.102
	5.0.101
	5.0.100
	5.0.100-rc.2.20479.15
	3.1.406
	3.1.405
	3.1.404
	3.1.403
	3.1.402
	3.1.401
	3.1.302
	3.1.301
	3.1.300
	3.1.201
	3.1.200
	3.1.102
	3.1.100
	3.1.100-preview3-014645
	3.1.100-preview2-014569
	3.0.100
	2.1.701
	2.1.700
	2.1.505
	2.1.504
	2.1.500
	2.1.402
	2.1.302
	2.1.301
	2.1.4
MSBuild SDKs: /Applications/Visual Studio.app/Contents/Resources/lib/monodevelop/bin/MSBuild/Current/bin/Sdks

=== .NET Core Runtime ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	5.0.3
	5.0.2
	5.0.1
	5.0.0
	5.0.0-rc.2.20475.5
	3.1.12
	3.1.11
	3.1.10
	3.1.9
	3.1.8
	3.1.7
	3.1.6
	3.1.5
	3.1.4
	3.1.3
	3.1.2
	3.1.0
	3.1.0-preview3.19553.2
	3.1.0-preview2.19525.6
	3.0.0
	2.1.23
	2.1.22
	2.1.21
	2.1.20
	2.1.19
	2.1.18
	2.1.17
	2.1.16
	2.1.15
	2.1.14
	2.1.13
	2.1.12
	2.1.11
	2.1.9
	2.1.8
	2.1.6
	2.1.4
	2.1.2
	2.1.1
	2.0.5

=== .NET Core 3.1 SDK ===

SDK: 3.1.406

=== Xamarin.Profiler ===

Version: 1.6.12.31
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Apple Developer Tools ===

Xcode 12.4 (17801)
Build 12D4e

=== Xamarin.Mac ===

Version: 7.8.1.0 (Visual Studio Enterprise)
Hash: 17845a189
Branch: d16-9
Build date: 2021-01-26 03:42:37-0500

=== Xamarin.iOS ===

Version: 14.14.1.0 (Visual Studio Enterprise)
Hash: 17845a189
Branch: d16-9
Build date: 2021-01-26 03:42:38-0500

=== Xamarin Designer ===

Version: 16.9.0.309
Hash: 9a5259fb5
Branch: remotes/origin/9a5259fb55e54a445c145ab5ea29c04a418deed2
Build date: 2021-02-04 09:34:28 UTC

=== Xamarin.Android ===

Version: 11.2.0.0 (Visual Studio Enterprise)
Commit: xamarin-android/d16-9/f908d16
Android SDK: /Users/astrakh/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		5.0 (API level 21)
		5.1 (API level 22)
		6.0 (API level 23)
		7.1 (API level 25)
		8.0 (API level 26)
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 29.0.4
SDK Build Tools Version: 29.0.2

Build Information: 
Mono: 5e9cb6d
Java.Interop: xamarin/java.interop/d16-9@1d382be
ProGuard: Guardsquare/proguard/v7.0.1@912d149
SQLite: xamarin/sqlite/3.32.2@cfe06e0
Xamarin.Android Tools: xamarin/xamarin-android-tools/main@ad80a42

=== Microsoft OpenJDK for Mobile ===

Java SDK: /Users/astrakh/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android SDK Manager ===

Version: 16.9.0.21
Hash: 57e40ba
Branch: remotes/origin/main
Build date: 2021-01-08 01:57:14 UTC

=== Android Device Manager ===

Version: 16.9.0.14
Hash: 0fdccda
Branch: remotes/origin/main
Build date: 2021-01-08 01:57:36 UTC

=== Build Information ===

Release ID: 809001593
Git revision: bf434e0df6f858da3d221515385acb8fe80b622a
Build date: 2021-02-11 10:01:48-05
Build branch: release-8.9
Xamarin extensions: bf434e0df6f858da3d221515385acb8fe80b622a

=== Operating System ===

Mac OS X 10.16.0
Darwin 20.2.0 Darwin Kernel Version 20.2.0
    Wed Dec  2 20:39:59 PST 2020
    root:xnu-7195.60.75~1/RELEASE_X86_64 x86_64

=== Enabled user installed extensions ===

XAML Styler 2.0.1
Xamarin Hot Reload 20190830.17


```

### Log File

[deployment-new.log](https://github.com/xamarin/xamarin-android/files/5991881/deployment-new.log)

**Comment by dellis1972** [comment]
Should be fixed by https://github.com/xamarin/monodroid/pull/1166

**Comment by TomBruyneel** [comment]
> Should be fixed by [xamarin/monodroid#1166](https://github.com/xamarin/monodroid/pull/1166)

Oh good, because I have the same problem with almost every nuget package I use. Your link is not working though?

**Comment by dellis1972** [comment]
The link is to a private repo. 
The best workaround is to upgrade your nuget package. The reason it happens is there are some old nugets still using the old pdb format. These have to be converted to mdb files for the debugger to use them. All of the new projects should be using the `portable` pdb format which the native runtime has built in support for. So if you can upgrade the problem packages.

**Comment by MagicAndre1981** [comment]
I already submitted a PR to Acr.UserDialogs to change PDB to portable and this is already merged. So use last nuget and be happy

**Comment by dellis1972** [comment]
A workaround for people who cannot get an updated nuget package to fix the issue

```xml
<Target Name="_RemoveMdbFiles" BeforeTargets="_Upload">
    <ItemGroup>
        <_ResolvedSymbols Remove="**/*.dll.mdb" />
    </ItemGroup>
</Target>
```

add this to the bottom of your `csproj` just above the `</Project>` element.

**Comment by dellis1972** [comment]
The fix for this was merged and included in this bump https://github.com/xamarin/xamarin-android/pull/5645. It should be in d16-10 P2. Please reopen if you find a case where none of the workarounds work.

**Comment by dileep-maxpro** [comment]
> A workaround for people who cannot get an updated nuget package to fix the issue
> 
> ```
> <Target Name="_RemoveMdbFiles" BeforeTargets="_Upload">
>     <ItemGroup>
>         <_ResolvedSymbols Remove="**/*.dll.mdb" />
>     </ItemGroup>
> </Target>
> ```
> 
> add this to the bottom of your `csproj` just above the `</Project>` element.

Tried this, but still getting the same issue. with Xamarin.Auth.dll.mdb

**Comment by MagicAndre1981** [comment]
> Tried this, but still getting the same issue. with Xamarin.Auth.dll.mdb

I think this lib is dead. Check if Xamarin.Essentials has the API you need or fork the lib and [change debug type on your own](https://github.com/xamarin/Xamarin.Auth/blob/46f959b45c8da4b3abdb6a4f0b57f66c6b77cc81/source/Core/Xamarin.Auth.XamarinAndroid/Xamarin.Auth.XamarinAndroid.csproj#L22)

**Comment by dellis1972** [comment]
@dileep-maxpro  do you have a diagnostic build log we can take a look at? The best way to get one is via 

```
msbuild <your Android csproj> /restore /t:Install /bl /p: _FastDeploymentDiagnosticLogging=true
```

That should produce a msbuild.binlog file. You can then zip this up and attach it to this Issue.

**Comment by louisoftokyo** [comment]
> A workaround for people who cannot get an updated nuget package to fix the issue
> 
> ```
> <Target Name="_RemoveMdbFiles" BeforeTargets="_Upload">
>     <ItemGroup>
>         <_ResolvedSymbols Remove="**/*.dll.mdb" />
>     </ItemGroup>
> </Target>
> ```
> 
> add this to the bottom of your `csproj` just above the `</Project>` element.

This unfortunately didn't work for me. It apparently compiles but I'm getting this when I attempt to deploy:

`Could not find file '<project path>\obj\Debug\100\android\assets\Microsoft.IdentityModel.Clients.ActiveDirectory.dll.mdb'.`

Attempting to run the above msbuild command failed with the same error.

**Comment by dan-SLT** [comment]
I am having the exact same issue as @louisoftokyo.

**Comment by dellis1972** [comment]
So the workaround target does work 

```
Removed Item(s): _ResolvedSymbols=obj\Debug\100\android\assets\Zambezi.dll.mdb
```

But the format of the command is missing files like `Microsoft.IdentityModel.Clients.ActiveDirectory.dll.mdb` because of the multiple periods. A new target which handles this scenario is below.


```
<Target Name="_RemoveMdbFiles" BeforeTargets="_Upload">
    <Message Text="%(_ResolvedSymbols.Extension)" />
    <ItemGroup>
        <_Temp Include="%(_ResolvedSymbols.Identity)" Condition="'%(_ResolvedSymbols.Extension)' == '.mdb'" />
        <_ResolvedSymbols Remove="@(_Temp)"/>
    </ItemGroup>
</Target>
```

**Comment by dan-SLT** [comment]
> So the workaround target does work
> 
> ```
> Removed Item(s): _ResolvedSymbols=obj\Debug\100\android\assets\Zambezi.dll.mdb
> ```
> 
> But the format of the command is missing files like `Microsoft.IdentityModel.Clients.ActiveDirectory.dll.mdb` because of the multiple periods. A new target which handles this scenario is below.
> 
> ```
> <Target Name="_RemoveMdbFiles" BeforeTargets="_Upload">
>     <Message Text="%(_ResolvedSymbols.Extension)" />
>     <ItemGroup>
>         <_Temp Include="%(_ResolvedSymbols.Identity)" Condition="'%(_ResolvedSymbols.Extension)' == '.mdb'" />
>         <_ResolvedSymbols Remove="@(_Temp)"/>
>     </ItemGroup>
> </Target>
> ```

I am unfortunately still getting the same error.

Severity	Code	Description	Project	File	Line	Suppression State
Error		Could not find file 'D:\Source\repos\SLT_APP\SLT_APP_TEST\SLT_APP.Android\obj\Debug\100\android\assets\Xamarin.Auth.dll.mdb'.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at Xamarin.Android.Tasks.FastDeploy.<DeployFileWithFastDevTool>d__103.MoveNext() in /Users/builder/azdo/_work/278/s/xamarin-android/external/monodroid/tools/msbuild/Tasks/FastDeploy.cs:line 621
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Xamarin.Android.Tasks.FastDeploy.<DeployFastDevFiles>d__102.MoveNext() in /Users/builder/azdo/_work/278/s/xamarin-android/external/monodroid/tools/msbuild/Tasks/FastDeploy.cs:line 598
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Xamarin.Android.Tasks.FastDeploy.<RunTaskAsync>d__90.MoveNext() in /Users/builder/azdo/_work/278/s/xamarin-android/external/monodroid/tools/msbuild/Tasks/FastDeploy.cs:line 223		 	0

**Comment by dellis1972** [comment]
@dan-SLT 

can you provide the diagnostic output using the command outlined in https://github.com/xamarin/xamarin-android/issues/5631#issuecomment-793708205.

**Comment by dan-SLT** [comment]
> @dan-SLT
> 
> can you provide the diagnostic output using the command outlined in [#5631 (comment)](https://github.com/xamarin/xamarin-android/issues/5631#issuecomment-793708205).

Running what you requested I get the errors below.

MSBUILD : error MSB1005: Specify a property and its value.
Switch: /p:

MSBUILD : error MSB1008: Only one project can be specified.
Switch: _FastDeploymentDiagnosticLogging=true

**Comment by dellis1972** [comment]
@dan-SLT 

There was a typo in the command there was a space between `/p:` and `_FastDeploymentDiagnosticLogging=true`.
The actual command is. 

```
msbuild <your Android csproj> /restore /t:Install /bl /p:_FastDeploymentDiagnosticLogging=true
```

**Comment by louisoftokyo** [comment]
> So the workaround target does work
> 
> ```
> Removed Item(s): _ResolvedSymbols=obj\Debug\100\android\assets\Zambezi.dll.mdb
> ```
> 
> But the format of the command is missing files like `Microsoft.IdentityModel.Clients.ActiveDirectory.dll.mdb` because of the multiple periods. A new target which handles this scenario is below.
> 
> ```
> <Target Name="_RemoveMdbFiles" BeforeTargets="_Upload">
>     <Message Text="%(_ResolvedSymbols.Extension)" />
>     <ItemGroup>
>         <_Temp Include="%(_ResolvedSymbols.Identity)" Condition="'%(_ResolvedSymbols.Extension)' == '.mdb'" />
>         <_ResolvedSymbols Remove="@(_Temp)"/>
>     </ItemGroup>
> </Target>
> ```

This now worked for me. Thank you very much.

**Comment by dileep-maxpro** [comment]
@dellis1972  Sure shall share when I run again.  for now I unchecked the fast deployment and continue working for now.

**Comment by dan-SLT** [comment]
> @dan-SLT
> 
> There was a typo in the command there was a space between `/p:` and `_FastDeploymentDiagnosticLogging=true`.
> The actual command is.
> 
> ```
> msbuild <your Android csproj> /restore /t:Install /bl /p:_FastDeploymentDiagnosticLogging=true
> ```

Getting this issue now. Not sure what changed.

Severity	Code	Description	Project	File	Line	Suppression State
Error		System.ArgumentException: Illegal characters in path.
   at System.IO.Path.CheckInvalidPathChars(String path, Boolean checkAdditional)
   at System.IO.Path.Combine(String path1, String path2)
   at Xamarin.Android.Tools.ProcessUtils.<FindExecutablesInDirectory>d__9.MoveNext() in /Users/builder/azdo/_work/278/s/xamarin-android/external/xamarin-android-tools/src/Xamarin.Android.Tools.AndroidSdk/ProcessUtils.cs:line 177
   at Xamarin.Android.Tools.ProcessUtils.<FindExecutablesInPath>d__8.MoveNext() in /Users/builder/azdo/_work/278/s/xamarin-android/external/xamarin-android-tools/src/Xamarin.Android.Tools.AndroidSdk/ProcessUtils.cs:line 168
   at Xamarin.Android.Tools.AndroidSdkBase.<GetAllAvailableAndroidNdks>d__73.MoveNext() in /Users/builder/azdo/_work/278/s/xamarin-android/external/xamarin-android-tools/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs:line 153
   at Xamarin.Android.Tools.AndroidSdkWindows.<GetAllAvailableAndroidNdks>d__43.MoveNext() in /Users/builder/azdo/_work/278/s/xamarin-android/external/xamarin-android-tools/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkWindows.cs:line 257
   at Xamarin.Android.Tools.AndroidSdkBase.GetValidNdkPath(String ctorParam) in /Users/builder/azdo/_work/278/s/xamarin-android/external/xamarin-android-tools/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs:line 128
   at Xamarin.Android.Tools.AndroidSdkBase.Initialize(String androidSdkPath, String androidNdkPath, String javaSdkPath) in /Users/builder/azdo/_work/278/s/xamarin-android/external/xamarin-android-tools/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs:line 71
   at Xamarin.Android.Tools.AndroidSdkWindows.Initialize(String androidSdkPath, String androidNdkPath, String javaSdkPath) in /Users/builder/azdo/_work/278/s/xamarin-android/external/xamarin-android-tools/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkWindows.cs:line 310
   at Xamarin.Android.Tools.AndroidSdkInfo..ctor(Action`2 logger, String androidSdkPath, String androidNdkPath, String javaSdkPath) in /Users/builder/azdo/_work/278/s/xamarin-android/external/xamarin-android-tools/src/Xamarin.Android.Tools.AndroidSdk/AndroidSdkInfo.cs:line 18
   at Xamarin.Android.Tasks.MonoAndroidHelper.RefreshAndroidSdk(String sdkPath, String ndkPath, String javaPath, TaskLoggingHelper logHelper)
   at Xamarin.Android.Tasks.ResolveSdks.RunTask()
   at Xamarin.Android.Tasks.AndroidTask.Execute()	SLT_APP.Android

**Comment by dan-SLT** [comment]
I reverted to Visual Studio 16.8.6 and re installed some Android SDK packages and it appears to be working.

**Comment by fa5tkill3r** [comment]
I have same issue as @dan-SLT

**Comment by Barokai** [comment]
had the same problems - were solved with updating Visual Studio 16.9.2 - no need for workarounds/adaptions of csproj anymore.
